### ISSUE TITLE
Add non-cluster host to Kubernets Calico network

### ISSUE THREAD

**Issue by AlejoAsd** [projectcalico/calico#3551]
## Context

I am trying to establish a connection between a docker container running on a non-cluster host and a pod inside a Kubernetes cluster running Calico. I have followed the instructions in [About non-cluster hosts](https://docs.projectcalico.org/getting-started/bare-metal/about) and [Docker container install](https://docs.projectcalico.org/getting-started/bare-metal/installation/container) to add the non-cluster host to the Calico network, but I run into problems when launching the `calico/node` container.

My setup consists of an unmanaged single-host Kubernetes cluster running in an AWS EC2 instance, and an Elastic Beanstalk (EB) docker application. Calico is running using Kubernetes as its store and has IP-in-IP encapsulation enabled for cross subnet traffic. The EB application creates a single EC2 instance that needs to communicate with several pods in the cluster. The application and the pods in the cluster are in the same AWS VPC, but may be in different subnets.

## Expected Behavior

Running a `calico/node` docker container in a non-cluster host should add the host to the Calico network.

## Current Behavior

I ran the following command in the EB application instance to launch the `calico/node` container and add the host to the Calico network:

```
 sudo docker run --rm --net=host --privileged \
  --name="calico-node" \
  -e DATASTORE_TYPE="kubernetes" \
  -e KUBECONFIG="/tmp/kube_config" \
  -e CALICO_NODENAME="test" \
  -e CALICO_NETWORKING_BACKEND="bird" \
  -v /var/log/calico:/var/log/calico \
  -v /run/docker/plugins:/run/docker/plugins \
  -v /lib/modules:/lib/modules \
  -v /var/run/calico:/var/run/calico \
  -v /etc/pki:/pki \
  -v /home/ec2-user/.kube/config:/tmp/kube_config \
  calico/node:v3.13.3
```

Which returned the following error message:

```
2020-05-13 22:47:33.819 [INFO][8] startup.go 309: Early log level set to info
2020-05-13 22:47:33.823 [INFO][8] startup.go 329: Using HOSTNAME environment (lowercase) for node name
2020-05-13 22:47:33.823 [INFO][8] startup.go 337: Determined node name: ip-172-30-39-148
2020-05-13 22:47:33.831 [INFO][8] startup.go 106: Skipping datastore connection test
2020-05-13 22:47:33.866 [INFO][8] startup.go 420: Building new node resource Name="ip-172-30-39-148"
2020-05-13 22:47:33.867 [INFO][8] startup.go 435: Initialize BGP data
2020-05-13 22:47:33.867 [INFO][8] startup.go 671: Using autodetected IPv4 address on interface eth0: 172.30.39.148/20
2020-05-13 22:47:33.868 [INFO][8] startup.go 505: Node IPv4 changed, will check for conflicts
2020-05-13 22:47:33.871 [INFO][8] startup.go 734: No AS number configured on node resource, using global value
2020-05-13 22:47:33.871 [WARNING][8] startup_linux.go 47: Expected /var/lib/calico to be mounted into the container but it wasn't present. Node name may not be detected properly
2020-05-13 22:47:33.875 [WARNING][8] node.go 64: Operation Create is not supported on Node type
2020-05-13 22:47:33.879 [ERROR][8] startup.go 217: Unable to set node resource configuration error=operation Create is not supported on Node(ip-A-B-C-D)
2020-05-13 22:47:33.879 [WARNING][8] startup.go 1214: Terminating
Calico node failed to start
```

A quick search led me to issues #3397 and #3407, which if I understood correctly indicate that my use case is not supported. Could you please confirm this? If that is the case, I am not sure I understand what use cases the `calico/node` docker installation is meant for.

I created this issue to confirm that I haven't missed anything and that this is the right way to go about things. 

If this is an issue of my specific setup, please let me know so that I can understand how to change it.

Thank you for your time!

## Possible Solution

Not a solution, but if my use case is indeed not supported, the documentation should be updated to reflect this.

## Steps to Reproduce (for bugs)

1. Create single-host Kubernetes cluster with Calico in an AWS EC2 instance with `kubeadm`.
2. Launch an additional AWS EC2 instance and try to run a `calico/node` container.

## Your Environment

* Calico version 3.13.3
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes 1.14.10
* Operating System and version: Amazon Linux AMI (Based on CentOS 6)
* Link to your project (optional):

**Comment by AlejoAsd** [comment]
Would changing the Calico store from Kubernetes to etcd allow adding the node to the network?

**Comment by AlejoAsd** [comment]
Ended up using Calico for policy only. Leaving this issue open to discuss documentation updates.

**Comment by caseydavenport** [comment]
@AlejoAsd sorry for delay on response here - agree we should make this work.

Using Calico in etcd mode should make this scenario work. In fact, this is the primary use-case for using etcd directly instead of Kubernetes mode. 

However, I think we should enhance Calico to make this work without the need for etcd mode. The main issue is that in k8s mode we assume all nodes are k8s nodes, and use the k8s node API to get node information. 

We'll need to either tweak the code to support running when a k8s node doesn't exist, or you could create dummy nodes in the k8s API to represent these external nodes.

**Comment by AlejoAsd** [comment]
After opening this issue and reading the documentation more thoroughly I found out that `etcd` indeed had support for creating nodes, while k8s didn't.

> The main issue is that in k8s mode we assume all nodes are k8s nodes, and use the k8s node API to get node information.

While it would certainly be nice to have k8s support, I think this is fine for now as you at least have a working alternative. 

To give this some closure, we should update the documentation to better reflect that k8s does not have support for nodes out of the cluster just yet. I can do that. 

I plan on editing the following pages:

* [About non-cluster hosts](https://docs.projectcalico.org/getting-started/bare-metal/about)
* [Docker container install](https://docs.projectcalico.org/getting-started/bare-metal/installation/container)

Let me know if there's anywhere else this should be mentioned.

**Comment by pshizhsysu** [comment]
1、create a virtual node in k8s

```
$ kubectl create -f - <<EOF
> apiVersion: v1
> kind: Node
> metadata:
>   name: peng03
> EOF 

$ kubectl get node
NAME     STATUS     ROLES    AGE   VERSION
peng01   Ready      master   61d   v1.17.3
peng02   Ready      <none>   26d   v1.17.3
peng03   NotReady   <none>   28s 
```

2、then install calico on non-cluster-host（docker mode）, it works

```
$ systemctl start calico

$ docker ps
CONTAINER ID        IMAGE                 COMMAND             CREATED             STATUS              PORTS               NAMES
75d0f627c632        calico/node:v3.16.1   "start_runit"       2 minutes ago      Up 2 minutes                           calico-node

$ route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         192.168.2.2     0.0.0.0         UG    100    0        0 ens33
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
172.26.14.192   192.168.2.102   255.255.255.192 UG    0      0        0 tunl0
172.26.71.192   192.168.2.101   255.255.255.192 UG    0      0        0 tunl0
172.26.176.0    0.0.0.0         255.255.255.192 U     0      0        0 *
192.168.2.0     0.0.0.0         255.255.255.0   U     100    0        0 ens33
```

**Comment by pankajagarwal** [comment]
I am running into the same issue although for me the calico node process and routes etc are populated fine (thanks to the instructions shared by @pshizhsysu ) but running a tracepath or curl to a running pod from this external host is still failing and timing out. 

What am i missing here or this documented steps are missing to allow an external host talk to a pod(s).

**Comment by ralfbecker** [comment]
Any progress / work on getting non-cluster hosts with datastore kubernetes working, WITHOUT the need to add a pseudo-node to K8s? I fear adding a node resource to K8s is not a good idea in a managed K8s setting ...

Ralf

**Comment by pankajagarwal** [comment]
a general note for other's running into this very issue, you can have a single external node connected to two separate Kubernetes cluster(s) due to limitation within Linux kernal code (there can only be one tunl0 interface for IpinIP). 

The other alternative useful in some cases might be to have a proxy service running in each cluster and have the request routed via the proxy to the pod's.

**Comment by ralfbecker** [comment]
I need to connect multiple external hosts to all hosts of a single K8s cluster using IPinIP and Wireguard.
Due to my managed K8s, which eg. rebuilds nodes automatic, I don't want to create fake node resources in the cluster.
I'm thinking about using a daemonSet to add a classic Wireguard config for each of these non-cluster hosts, and something similar on the non-cluster host to fetch the public keys of the nodes from the Calico config/CRD.

**Comment by ralfbecker** [comment]
One simple way to archive this would be, if Felix checks the AllowedIPs of peers against the cluster CIDR.
K8s nodes match the CIDR can should get cleaned up, but non-cluster hosts would be left alone.

Any way to get some Felix developers into this thread, or do I need to open a bug report in the Felix repo?

**Comment by caseydavenport** [comment]
I think ultimately we need a way to represent the node in the k8s API for Calico to understand that it is valid. I don't believe a CIDR-based approach will work for this, not all networks are built such that using CIDRs to identify nodes is feasible. 

We can either use the existing Kubernetes Node API, or add a new `ExternalNode` CRD that Calico can manage itself. I am leaning towards the latter, but it needs extensive design review and examination of use-cases before we can move forward.

**Comment by ralfbecker** [comment]
I also prefer an ExternalNode CRD, as in a managed Kubernetes cluster with automatic rebuild of the nodes, I don't want to screw up the process by creating fake nodes. 

Use case for us are NFS- and mail-backends, which use ZFS as filesystem with huge storage pools, not possible in the (manage) K8s cluster. It would also allow us to move these hosts to other cloud-providers or use own servers in a co-location.

**Comment by KlavsKlavsen** [comment]
Would this work, to connect 2 (or multiple) k8s clusters, which all run calico ?

**Comment by muvaf** [comment]
We just released [`limguard`](https://github.com/limrun-inc/limguard) project that separates WireGuard from CNI to provide a LAN-like layer to Calico and others, allowing any Linux and macOS hosts to join the same network.

One thing you'd still need to do is to participate in Calico BGP for routes to be registered or manually add the CIDRs via `ip route` (Linux) or `route -n add` (macOS). This is taken care of by Felix, e.g. `calico-node` pods, but something you need to do if nodes aren't running that. `limguard` scans all the routes continuously and detect the ones targeted for its peer IPs and add them to Wireguard's `allowed-ips` list of that given peer, which is a requirement of WireGuard.

We built it for our clusters comprising of private/public Linux and macOS (via virtual-kubelet) nodes and is in production at [lim.run](https://lim.run)
